File: ofdm_phy/link.py
"""Link adaptation: rate ladder, link-control word, per-direction controller.

Two half-duplex stations form two directed links (Tx1->Rx2, Tx2->Rx1), each
governed by its RECEIVER: only the receiver knows its own noise floor, so it
requests the transmission rung, and the request rides in every frame's
20-bit link-control word. Modes are self-labeling at the PHY (per-mode ZC
preamble roots + header mod/spd fields), so a mode switch can never deadlock
the link -- worst case a frame is lost, the transmitter's loss fallback
kicks in, and both sides re-converge at rung 0 (EXTREME), which works
whenever anything works.

Link-control word (packed into Data.reserved, 20 bits):
    seq(2) | ack(2) | req_rung(4) | snr_report(4) | freq_corr(5) | flags(3)
SNR report: 2 dB steps over -24 .. +6 dB. freq_corr: the AFC/netting request
"shift your carrier by this many Hz", signed, 8 Hz steps, +-120 Hz per frame
(larger offsets converge iteratively). Stop-and-wait ARQ needs only 2-bit
sequence numbers.
"""

import collections
import typing
from dataclasses import dataclass, field

from .modes import LinkMode
from .packets import ModType, CCSpeed


@dataclass(frozen=True)
class Rung:
    mode: LinkMode
    mod: ModType
    spd: CCSpeed
    sens_db: float      # PER<=10% sensitivity, article channel, recal RX
    user_rate: float    # bit/s after FEC
    measured: bool      # False = interpolated from neighbouring measurements


# Dominated rungs (slower AND less sensitive than a neighbour) are omitted,
# e.g. NORMAL BPSK 2/3 (-5.2 dB, 235 bit/s) loses to NORMAL QPSK 1/3
# (-5.6 dB, 235 bit/s).
LADDER = [
    Rung(LinkMode.EXTREME, ModType.BPSK, CCSpeed.R13, -17.9, 7.8, True),
    Rung(LinkMode.ROBUST, ModType.BPSK, CCSpeed.R13, -11.7, 30.8, True),
    Rung(LinkMode.ROBUST, ModType.BPSK, CCSpeed.R12, -11.8, 46.2, True),
    Rung(LinkMode.ROBUST, ModType.QPSK, CCSpeed.R13, -11.3, 61.5, True),
    Rung(LinkMode.NORMAL, ModType.BPSK, CCSpeed.R13, -7.6, 117.6, True),
    Rung(LinkMode.NORMAL, ModType.BPSK, CCSpeed.R12, -7.3, 176.5, True),
    Rung(LinkMode.NORMAL, ModType.QPSK, CCSpeed.R13, -7.0, 235.3, True),
    Rung(LinkMode.NORMAL, ModType.QPSK, CCSpeed.R12, -5.3, 352.9, True),
    Rung(LinkMode.NORMAL, ModType.QPSK, CCSpeed.R23, -3.8, 470.6, True),
    Rung(LinkMode.NORMAL, ModType.QPSK, CCSpeed.R34, -2.2, 529.4, True),
    Rung(LinkMode.NORMAL, ModType.QAM16, CCSpeed.R12, 0.7, 705.9, True),
    Rung(LinkMode.NORMAL, ModType.QAM16, CCSpeed.R23, 2.6, 941.2, True),
    Rung(LinkMode.NORMAL, ModType.QAM16, CCSpeed.R34, 4.7, 1058.8, True),
]


# --- link-control word ------------------------------------------------------

FREQ_STEP_HZ = 8.0
FREQ_MAX_HZ = 15 * FREQ_STEP_HZ  # +-120 Hz per frame

# The 4-bit SNR field's code 0 means "I have no measurement to report",
# not "-24 dB". Reporting a measurement one does not have is not a
# harmless white lie: the peer caps its transmit rung on that number, so
# a station that has simply been quiet for a minute used to tell its
# peer "I hear you at -24 dB" and get 19-second EXTREME acks back on a
# +20 dB link (measured on the two-board stand, 3 of 4 replies).
# Genuine measurements therefore clamp into codes 1..15 = -22..+6 dB;
# anything at or below -22 dB caps the peer to rung 0 either way, so the
# lost code costs nothing. Old peers pack their "no measurement"
# sentinel into this same code 0, so a patched receiver reads them
# correctly too, and an unpatched receiver reads code 0 as -24 dB
# exactly as before.
LC_SNR_NONE = -99.0
LC_SNR_NONE_MAX = -90.0    # anything this low means "no measurement"


@dataclass
class LinkControl:
    seq: int = 0
    ack: int = 0
    req_rung: int = 0
    snr_db: float = LC_SNR_NONE   # nothing measured, until one is
    freq_corr_hz: float = 0.0  # "peer, shift your carrier by this much"
    flags: int = 0

    def pack(self) -> int:
        if self.snr_db <= LC_SNR_NONE_MAX:
            snr_q = 0                       # no measurement to report
        else:                               # a real one: never code 0
            snr_q = max(1, min(15, int(round((self.snr_db + 24.0) / 2))))
        f_q = max(-15, min(15, int(round(self.freq_corr_hz / FREQ_STEP_HZ)))) + 15
        return ((self.seq & 3) << 18) | ((self.ack & 3) << 16) | \
               ((self.req_rung & 15) << 12) | (snr_q << 8) | \
               (f_q << 3) | (self.flags & 7)

    @classmethod
    def unpack(cls, v: int) -> "LinkControl":
        return cls(seq=(v >> 18) & 3, ack=(v >> 16) & 3,
                   req_rung=(v >> 12) & 15,
                   snr_db=(LC_SNR_NONE if ((v >> 8) & 15) == 0
                           else ((v >> 8) & 15) * 2.0 - 24.0),
                   freq_corr_hz=(((v >> 3) & 31) - 15) * FREQ_STEP_HZ,
                   flags=v & 7)


# --- controller -------------------------------------------------------------

@dataclass
class LinkController:
    """One per station; manages the receive-side request for the inbound
    link and the transmit-side rung choice for the outbound link."""

    margin_up: float = 2.5      # dB above rung sensitivity to move up
    margin_keep: float = 1.0    # dB below which the request drops
    stale_s: float = 90.0       # peer-request decay interval
    rx_stale_s: float = 45.0    # inbound-silence decay interval
    history_len: int = 5

    # RX side (inbound link)
    _snr_hist: typing.Deque = field(default_factory=lambda: collections.deque(maxlen=5))
    _my_req: int = 0
    last_rx_time: float = -1e9
    # how much of the inbound silence has already been charged to the
    # request: the decay must follow elapsed time, not the number of
    # times it is asked (the C station asks twice per frame)
    _req_decay_t: float = -1e9

    # TX side (outbound link)
    peer_req: int = 0
    peer_req_time: float = -1e9
    peer_report_db: float = -99.0  # peer's measured SNR of MY signal
    consecutive_losses: int = 0
    # learned per-rung sensitivity corrections: the static table comes from
    # one channel model; real deployments differ, so every ack/timeout
    # nudges the effective sensitivity of the rung that was used
    rung_offset_db: list = field(default_factory=lambda: [0.0] * len(LADDER))

    # --- receive side ------------------------------------------------------

    snr_max_age_s: float = 60.0

    def on_rx_frame(self, snr_db: float, lc: LinkControl, now: float):
        """Called for every successfully decoded inbound frame."""
        self._snr_hist.append((now, snr_db))
        self.last_rx_time = now
        self.peer_req = lc.req_rung
        self.peer_req_time = now
        self.peer_report_db = lc.snr_db
        self.consecutive_losses = 0

    def filtered_snr(self, now: float = None) -> float:
        """Fade-aware statistic: second-lowest of the recent, non-expired
        history (a low percentile, not the mean -- QSB punishes averages).
        Age-windowing matters: pre-fade measurements must not resurrect a
        high request after a dropout."""
        vals = [s for (t, s) in self._snr_hist
                if now is None or now - t <= self.snr_max_age_s]
        if not vals:
            return -99.0
        h = sorted(vals)
        return h[1] if len(h) >= 3 else h[0]

    def rx_request(self, now: float = None) -> int:
        """The rung I ask the peer to transmit at (with hysteresis).

        Inbound silence decays the request: measurements go stale the moment
        frames stop decoding, yet the request still travels in our own
        (possibly still-working) outbound frames with a fresh timestamp -- so
        the receiver itself must treat "I hear nothing" as evidence."""
        if now is not None:
            # Decay follows ELAPSED TIME, not call count. Subtracting the
            # whole silence on every call while leaving last_rx_time alone
            # meant the C twin -- which asks twice per frame -- walked the
            # request 12 -> 11 -> 10 -> 9 in three calls at one instant,
            # collapsing it to rung 0 after a handful of retransmissions.
            base = max(self._req_decay_t, self.last_rx_time)
            if now - base > self.rx_stale_s:
                decay = int((now - base) // self.rx_stale_s)
                self._my_req = max(0, self._my_req - decay)
                self._req_decay_t = base + decay * self.rx_stale_s
                self._snr_hist.clear()  # stale measurements must not return
                return self._my_req
            if now - self.last_rx_time > self.rx_stale_s:
                return self._my_req     # this silence is already charged

        snr = self.filtered_snr(now)
        best_up = 0
        for i, r in enumerate(LADDER):
            if snr >= r.sens_db + self.margin_up:
                best_up = i
        cur = LADDER[self._my_req]
        if best_up > self._my_req:
            self._my_req = best_up           # fast start / upshift
        elif snr < cur.sens_db + self.margin_keep:
            down = 0
            for i, r in enumerate(LADDER):   # highest rung still holding margin
                if snr >= r.sens_db + self.margin_keep:
                    down = i
            self._my_req = min(self._my_req, down)
        return self._my_req

    # --- transmit side -----------------------------------------------------

    def on_ack(self):
        self.consecutive_losses = 0

    def on_timeout(self):
        self.consecutive_losses += 1

    def note_outcome(self, rung: int, ok: bool):
        """Learn the deployment's real sensitivities: a loss at a rung makes
        it effectively less sensitive (+0.7 dB, asymmetric fast-up/slow-down),
        a success slowly earns the margin back (-0.15 dB)."""
        if ok:
            self.rung_offset_db[rung] = max(0.0, self.rung_offset_db[rung] - 0.15)
        else:
            self.rung_offset_db[rung] = min(6.0, self.rung_offset_db[rung] + 0.7)

    def tx_rung(self, now: float) -> int:
        """Rung for my next outbound frame: the peer's request, capped by the
        peer's SNR report of my signal, decayed by staleness, and cut by the
        loss-fallback ladder.

        The cap applies only when the peer actually reported a measurement.
        "I heard you badly" and "I have not heard anything lately" are
        different facts: the first must slow us down, the second is what
        every gap in a conversation looks like, and treating it as a -24 dB
        report collapsed the reply to EXTREME. A peer that has genuinely
        gone away is still handled -- by the staleness decay below, which
        walks the rung down one step per stale_s."""
        rung = self.peer_req

        if self.peer_report_db > LC_SNR_NONE_MAX:
            cap = 0
            for i, r in enumerate(LADDER):
                if self.peer_report_db >= r.sens_db + self.rung_offset_db[i] \
                        + self.margin_keep:
                    cap = i
            rung = min(rung, cap)

        age = now - self.peer_req_time
        if age > self.stale_s:
            rung = max(0, rung - int(age // self.stale_s))
        if self.consecutive_losses >= 4:
            rung = 0
        elif self.consecutive_losses >= 2:
            rung = max(0, rung - 2)
        return rung

    def tx_rung_for_class(self, now: float, qos: str = "bulk") -> int:
        """QoS classes are margin policy: control/interactive traffic rides
        one rung below the bulk choice (extra margin, shorter frames)."""
        rung = self.tx_rung(now)
        if qos in ("control", "interactive"):
            rung = max(0, rung - 1)
        return rung


def max_payload_bytes(rung: Rung, max_air_s: float = 8.0) -> int:
    """Air-time budget -> payload cap (latency QoS lever); the packet format
    itself caps payloads at 27 bytes."""
    return max(1, min(27, int(rung.user_rate * max_air_s / 8)))

File: cport/src/link.c
#include <math.h>
#include <string.h>

#include "link.h"
#include "rom_link.h"

#define MARGIN_UP 2.5
#define MARGIN_KEEP 1.0
#define STALE_S 90.0
#define RX_STALE_S 45.0
#define SNR_MAX_AGE_S 60.0

int ladder_n(void) { return LADDER_N; }
link_mode_t ladder_mode(int r) { return (link_mode_t)LADDER_MODE[r]; }
mod_type_t ladder_mod(int r) { return (mod_type_t)LADDER_MOD[r]; }
cc_rate_t ladder_spd(int r) { return (cc_rate_t)LADDER_SPD[r]; }
double ladder_sens_db(int r) { return LADDER_SENS[r]; }
double ladder_rate(int r) { return LADDER_RATE[r]; }

/* python round() = round-half-even = C rint() in the default mode */
uint32_t lc_pack(const lc_word_t *lc)
{
    int snr_q = LC_SNR_IS_NONE(lc->snr_db)
                    ? 0                       /* no measurement to report */
                    : (int)rint((lc->snr_db + 24.0) / 2.0);
    int f_q = (int)rint(lc->freq_corr_hz / FREQ_STEP_HZ);
    if (snr_q < 1 && !LC_SNR_IS_NONE(lc->snr_db))
        snr_q = 1;                            /* a real one is never code 0 */
    if (snr_q > 15) snr_q = 15;
    if (f_q < -15) f_q = -15;
    if (f_q > 15) f_q = 15;
    f_q += 15;
    return ((uint32_t)(lc->seq & 3) << 18) | ((uint32_t)(lc->ack & 3) << 16)
           | ((uint32_t)(lc->req_rung & 15) << 12)
           | ((uint32_t)snr_q << 8) | ((uint32_t)f_q << 3)
           | (uint32_t)(lc->flags & 7);
}

void lc_unpack(uint32_t v, lc_word_t *lc)
{
    lc->seq = (int)((v >> 18) & 3);
    lc->ack = (int)((v >> 16) & 3);
    lc->req_rung = (int)((v >> 12) & 15);
    lc->snr_db = ((v >> 8) & 15) == 0
                     ? LC_SNR_NONE
                     : (double)((v >> 8) & 15) * 2.0 - 24.0;
    lc->freq_corr_hz = (double)((int)((v >> 3) & 31) - 15) * FREQ_STEP_HZ;
    lc->flags = (int)(v & 7);
}

void ctl_init(link_ctl_t *c)
{
    memset(c, 0, sizeof(*c));
    c->last_rx_time = -1e9;
    c->peer_req_time = -1e9;
    c->peer_report_db = LC_SNR_NONE;
    c->req_decay_t = -1e9;
}

void ctl_on_rx_frame(link_ctl_t *c, double snr_db, const lc_word_t *lc,
                     double now)
{
    /* deque(maxlen=5): ring overwrite */
    int slot = (c->snr_hist_head + c->snr_hist_n) % SNR_HIST_LEN;
    if (c->snr_hist_n == SNR_HIST_LEN)
        c->snr_hist_head = (c->snr_hist_head + 1) % SNR_HIST_LEN;
    else
        c->snr_hist_n++;
    c->snr_hist_t[slot] = now;
    c->snr_hist_v[slot] = snr_db;
    c->last_rx_time = now;
    c->peer_req = lc->req_rung;
    c->peer_req_time = now;
    c->peer_report_db = lc->snr_db;
    c->consecutive_losses = 0;
}

double ctl_filtered_snr(const link_ctl_t *c, double now)
{
    double vals[SNR_HIST_LEN];
    int n = 0, i, j;
    for (i = 0; i < c->snr_hist_n; i++) {
        int s = (c->snr_hist_head + i) % SNR_HIST_LEN;
        if (now - c->snr_hist_t[s] <= SNR_MAX_AGE_S)
            vals[n++] = c->snr_hist_v[s];
    }
    if (n == 0)
        return -99.0;
    for (i = 1; i < n; i++) { /* insertion sort ascending */
        double v = vals[i];
        for (j = i - 1; j >= 0 && vals[j] > v; j--)
            vals[j + 1] = vals[j];
        vals[j + 1] = v;
    }
    return n >= 3 ? vals[1] : vals[0];
}

int ctl_rx_request(link_ctl_t *c, double now)
{
    double snr;
    int i, best_up = 0;

    /* Decay follows ELAPSED TIME, not the number of calls. This used
     * to subtract the whole silence every time while leaving
     * last_rx_time alone, and station_poll_tx asks twice per frame (the
     * request that goes on the wire, then the reply-rung guess):
     * measured, three calls at ONE instant walked my_req 12 -> 11 ->
     * 10 -> 9, so a few retransmissions inside a fade collapsed the
     * request to rung 0 and the peer was asked for EXTREME. Charging
     * the decay against req_decay_t makes repeated calls idempotent. */
    {
        double base = c->req_decay_t > c->last_rx_time ? c->req_decay_t
                                                       : c->last_rx_time;
        if (now - base > RX_STALE_S) {
            int decay = (int)floor((now - base) / RX_STALE_S);
            c->my_req -= decay;
            if (c->my_req < 0)
                c->my_req = 0;
            c->req_decay_t = base + decay * RX_STALE_S;
            c->snr_hist_n = 0; /* stale measurements must not come back */
            c->snr_hist_head = 0;
            return c->my_req;
        }
        if (now - c->last_rx_time > RX_STALE_S)
            return c->my_req;   /* this silence is already charged */
    }

    snr = ctl_filtered_snr(c, now);
    for (i = 0; i < LADDER_N; i++)
        if (snr >= LADDER_SENS[i] + MARGIN_UP)
            best_up = i;
    if (best_up > c->my_req) {
        c->my_req = best_up;
    } else if (snr < LADDER_SENS[c->my_req] + MARGIN_KEEP) {
        int down = 0;
        for (i = 0; i < LADDER_N; i++)
            if (snr >= LADDER_SENS[i] + MARGIN_KEEP)
                down = i;
        if (down < c->my_req)
            c->my_req = down;
    }
    return c->my_req;
}

void ctl_on_ack(link_ctl_t *c) { c->consecutive_losses = 0; }
void ctl_on_timeout(link_ctl_t *c) { c->consecutive_losses++; }

void ctl_note_outcome(link_ctl_t *c, int rung, int ok)
{
    if (ok) {
        c->rung_offset_db[rung] -= 0.15;
        if (c->rung_offset_db[rung] < 0.0)
            c->rung_offset_db[rung] = 0.0;
    } else {
        c->rung_offset_db[rung] += 0.7;
        if (c->rung_offset_db[rung] > 6.0)
            c->rung_offset_db[rung] = 6.0;
    }
}

int ctl_tx_rung(const link_ctl_t *c, double now)
{
    int rung = c->peer_req, cap = 0, i;
    double age;

    /* The cap applies only to a report the peer actually made. "I heard
     * you badly" and "I have not heard anything lately" are different
     * facts: the first must slow us down, the second is what every gap
     * in a conversation looks like. A peer that has really gone away is
     * still caught by the staleness decay below. */
    if (!LC_SNR_IS_NONE(c->peer_report_db)) {
        for (i = 0; i < LADDER_N; i++)
            if (c->peer_report_db
                >= LADDER_SENS[i] + c->rung_offset_db[i] + MARGIN_KEEP)
                cap = i;
        if (cap < rung)
            rung = cap;
    }

    age = now - c->peer_req_time;
    if (age > STALE_S) {
        rung -= (int)floor(age / STALE_S);
        if (rung < 0)
            rung = 0;
    }
    if (c->consecutive_losses >= 4)
        rung = 0;
    else if (c->consecutive_losses >= 2) {
        rung -= 2;
        if (rung < 0)
            rung = 0;
    }
    return rung;
}

void ctl_diag(const link_ctl_t *c, double now, link_diag_t *d)
{
    int i;
    d->rung = ctl_tx_rung(c, now);
    d->peer_req = c->peer_req;
    /* -1 = the peer reported no measurement, so no cap applies. It is
     * NOT the same as a cap of 0, and printing it as one is what made
     * this whole class of defect hard to read in the first place. */
    d->cap = -1;
    if (!LC_SNR_IS_NONE(c->peer_report_db)) {
        d->cap = 0;
        for (i = 0; i < LADDER_N; i++)
            if (c->peer_report_db
                >= LADDER_SENS[i] + c->rung_offset_db[i] + MARGIN_KEEP)
                d->cap = i;
    }
    d->losses = c->consecutive_losses;
    d->my_req = c->my_req;
    d->peer_report_db = c->peer_report_db;
    d->filtered_snr = ctl_filtered_snr(c, now);
    d->req_age_s = now - c->peer_req_time;
    d->rx_age_s = now - c->last_rx_time;
    d->offset_db = c->rung_offset_db[d->rung];
}

int ctl_tx_rung_for_class(const link_ctl_t *c, double now, int qos)
{
    int rung = ctl_tx_rung(c, now);
    if (qos <= 1) { /* control / interactive */
        rung -= 1;
        if (rung < 0)
            rung = 0;
    }
    return rung;
}

int link_max_payload_bytes(int rung, double max_air_s)
{
    int v = (int)(LADDER_RATE[rung] * max_air_s / 8.0);
    if (v > 27)
        v = 27;
    if (v < 1)
        v = 1;
    return v;
}
